design-documents/node-static-content-deploy.md

```diff
@@ -0,0 +1,95 @@
+# Modernize Static Content Deployment
+
+**Note: This is a proposal for a possible improvement in the future. Even if we reach consensus, there are no guarantees yet that this would be implemented. Exploratory for the moment**
+
+- [Static Content Deployment](#static-content-deployment)
+  * [Docs](#docs)
+  * [Use Cases](#use-cases)
+  * [What does it do?](#what-does-it-do-)
+  * [Problems with Static Content Deployment](#problems-with-static-content-deployment)
+  * [Why rewrite Static Content Deployment?](#why-rewrite-static-content-deployment-)
+    + [Performance](#performance)
+    + [Access to Modern Tooling](#access-to-modern-tooling)
+    + [Accessible for Front-End Devs](#accessible-for-front-end-devs)
+    + [What opportunities open up with a re-write of SCD using node.js?](#what-opportunities-open-up-with-a-re-write-of-scd-using-nodejs-)
+  * [How do you re-write Static Content Deployment?](#how-do-you-re-write-static-content-deployment-)
+  
+## Docs
+- [Deploying Static View Files](https://devdocs.magento.com/guides/v2.3/config-guide/cli/config-cli-subcommands-static-view.html)
+- [Strategies](https://devdocs.magento.com/guides/v2.3/config-guide/cli/config-cli-subcommands-static-deploy-strategies.html)
+
+## Use Cases
+1. Pre-processing and flattening of static assets in local development (i.e. less)
+1. Pre-processing and flattening of static assets for production deployment
+
+## What does it do?
+
+- [Theme/module/i18n/"lib/web" file inheritance flattening](https://devdocs.magento.com/guides/v2.3/frontend-dev-guide/themes/theme-inherit.html)
+- Locale splitting
+    - Each locale gets its own static assets under a theme dir
+- Less compilation
```

Comment:
I agree about Less compilation and JS minification but not sure about other tasks like moving files, concatenation, and locales generation. I suppose the problem that you need to interact with the file system frequently and as node.js is also single threaded will it be a big difference between node.js and PHP? From my experience python or Go are more suitable for these tasks.

Reply:
> I suppose the problem that you need to interact with the file system frequently and as node.js is also single threaded will it be a big difference between node.js and PHP

An extremely large different. PHP blocks when reading a file, node does not.

node's original selling point was async network and disk I/O - that's where it thrives. Since more of static content deploy is copying/resolving/moving, rather than pure CPU-bound tasks, it's a good fit

Reply:
> From my experience python or Go are more suitable for these tasks.

This leads us back to the same problem as PHP (outlined in this document): lack of a good interop story with node. All modern front-end tooling is written assuming node as a runtime.

Reply:
It would be great to thing about just styles compilation and be able to extend those tools with other compilers like SASS, Stylus or some PostCSS based process.
I thing people will be happy to migrate from Frontools to offical tools 😊